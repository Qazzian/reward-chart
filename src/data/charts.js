import { combineReducers } from 'redux';

import {chart} from './chart';


const chartActionTypes = {
	CHART_ADD: 'CHART_ADD',
	CHART_REMOVE: 'CHART_REMOVE',
	CHART_ADD_EMOTE: 'CHART_ADD_EMOTE',
	CHART_REMOVE_EMOTE: 'CHART_ADD_EMOTE',
};

function validateAction(action) {
	if (!action.id) {
		throw new Error(`Action doesn't have a valid id`);
	}
}

// State reducers
const chartsById = (state = {}, action = {}) => {
	switch (action.type) {
		case chartActionTypes.CHART_ADD:
			validateAction(action);
			if (chartExists(state, action.id)) {
				throw new Error(`Chart ID '${action.id}' already exists`);
			}
			return {
				...state,
				[action.id]: chart(state[action.id], action),
			};
		case chartActionTypes.CHART_REMOVE:
			validateAction(action);
			return Object.keys(state).reduce((newState, chartId) => {
				if (parseInt(chartId, 10) !== action.id) {
					newState[chartId] = state[chartId];
				}
				return newState;
			}, {});
		case chartActionTypes.CHART_ADD_EMOTE:
		case chartActionTypes.CHART_REMOVE_EMOTE:
			return {
				...state,
				[action.chartId]: chart(state[action.chartId], action),
			};
		default:
			return state;
	}
};

const allIds = (state=[], action={}) => {
	switch(action.type) {
		case chartActionTypes.CHART_ADD:
			validateAction(action);
			return [
				...state,
				action.id,
			];
		case chartActionTypes.CHART_REMOVE:
			validateAction(action);
			return state.filter((chartId) => {
				return chartId !== action.id;
			});
		default:
			return state;
	}
};

const DEFAULT_STATE = {
	allIds: allIds(),
	chartsById: chartsById(),
	nextId: 1,
};


export const charts = (state, action) => {
	let newAction = {...action};
	let newState = {
		...DEFAULT_STATE,
		...state,
	};

	if (action && action.type === chartActionTypes.CHART_ADD && !action.id) {
		newAction.id = newState.nextId;
		newState.nextId += 1;
	}

	return {
		...newState,
		chartsById: chartsById(newState.chartsById, action && newAction),
		allIds: allIds(newState.allIds, action && newAction),
	};
};


// State queries
export const getAllCharts = (state) => {
	if (!state || !state.allIds.length) {
		return [];
	}
	const allCharts = state.allIds.map(id => {
		return state.chartsById[id];
	});
	return allCharts;
};

export function chartExists(state, chartId) {
	return state.chartsById && state.chartsById.hasOwnProperty(chartId);
}

export const getChartById = (state, id) => {
	if (!state || !state.chartsById) {
		return {};
	}
	return state.chartsById[id];
};


// State Actions
export const addChart = ({id, name}) => {
	return {
		type: chartActionTypes.CHART_ADD,
		name,
		id,
	};
};

export const removeChart = (id) => ({
	type: chartActionTypes.CHART_REMOVE,
	id,
});

export const chartAddEmote = (chartId, emoteId) => ({
	type: chartActionTypes.CHART_ADD_EMOTE,
	chartId,
	emoteId,
});

export const chartRemoveEmote = (chartId, emoteId) => ({
	type: chartActionTypes.CHART_REMOVE_EMOTE,
	chartId,
	emoteId,
});
