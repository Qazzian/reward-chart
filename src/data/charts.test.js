import * as fromCharts from "./charts";
import clock from "jest-plugin-clock";

const charts = fromCharts.charts;

describe('charts state', function () {


	it('should be defined', (done) => {
		expect(charts).toBeDefined();
		done();
	});

	it('should return an empty chart list by default', (done) => {
		expect(charts()).toMatchSnapshot();
		done();
	});

	it('should add a Chart to the list', (done) => {
		const firstState = charts([], fromCharts.addChart({name: 'test1'}));
		const chartList1 = fromCharts.getAllCharts(firstState);
		expect(chartList1.length).toBe(1);
		expect(chartList1[0].name).toBe('test1');


		const secondState = charts(firstState, fromCharts.addChart({name: 987}));
		console.info('ChartList:', secondState);

		expect(secondState).toMatchSnapshot();
		const chartList2 = fromCharts.getAllCharts(secondState);
		expect(chartList2.length).toBe(2);
		expect(chartList2[1].name).toBe(987);

		expect(chartList1.length).toBe(1);
		done();
	});

	it('should Remove a chart from the list', (done) => {
		const firstState = charts([], fromCharts.addChart({name: 123}));
		const secondState = charts(firstState, fromCharts.addChart({name: 987}));
		expect(fromCharts.getAllCharts(secondState).length).toBe(2);

		const thirdState = charts(secondState, fromCharts.removeChart(2));
		expect(thirdState).toMatchSnapshot();
		expect(fromCharts.getAllCharts(thirdState).length).toBe(1);
		done();
	});

	it('should Add and remove emotes to a chart', (done) => {
		clock.set('2018-12-14');
		const firstState = charts([], fromCharts.addChart({name: 123}));
		const secondState = charts(firstState, fromCharts.chartAddEmote(1, 23));
		expect(secondState).toMatchSnapshot();
		const thirdState = charts(secondState, fromCharts.chartAddEmote(1, 45));
		expect(thirdState).toMatchSnapshot();
		const fourthState = charts(secondState, fromCharts.chartRemoveEmote(1, 45));
		expect(fourthState).toMatchSnapshot();
		done();
	});

	xit('should add multiple emotes to a chart', () => {
		const firstState = charts([], charts.addChart({name: 123}));
		const secondState = charts(firstState, charts.chartHappy({name: 123}, new Date('2018-12-14')));
		const thirdState = charts(secondState, charts.chartSad({name: 123}, new Date('2018-12-15')));
		const forthState = charts(thirdState, charts.chartHappy({name: 123}, new Date('2018-12-16')));
		expect(forthState).toMatchSnapshot();
	});
});
