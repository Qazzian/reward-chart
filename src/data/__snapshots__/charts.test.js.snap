// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`charts state should Add and remove emotes to a chart 1`] = `
Object {
  "allIds": Array [
    1,
  ],
  "chartsById": Object {
    "1": Object {
      "emotes": Array [
        23,
      ],
      "id": 1,
      "name": 123,
    },
  },
  "nextId": 2,
}
`;

exports[`charts state should Add and remove emotes to a chart 2`] = `
Object {
  "allIds": Array [
    1,
  ],
  "chartsById": Object {
    "1": Object {
      "emotes": Array [
        23,
        45,
      ],
      "id": 1,
      "name": 123,
    },
  },
  "nextId": 2,
}
`;

exports[`charts state should Add and remove emotes to a chart 3`] = `
Object {
  "allIds": Array [
    1,
  ],
  "chartsById": Object {
    "1": Object {
      "emotes": Array [
        23,
        45,
      ],
      "id": 1,
      "name": 123,
    },
  },
  "nextId": 2,
}
`;

exports[`charts state should Remove a chart from the list 1`] = `
Object {
  "allIds": Array [
    1,
  ],
  "chartsById": Object {
    "1": Object {
      "emotes": Array [],
      "id": 1,
      "name": 123,
    },
  },
  "nextId": 3,
}
`;

exports[`charts state should add a Chart to the list 1`] = `
Object {
  "allIds": Array [
    1,
    2,
  ],
  "chartsById": Object {
    "1": Object {
      "emotes": Array [],
      "id": 1,
      "name": "test1",
    },
    "2": Object {
      "emotes": Array [],
      "id": 2,
      "name": 987,
    },
  },
  "nextId": 3,
}
`;

exports[`charts state should return an empty chart list by default 1`] = `
Object {
  "allIds": Array [],
  "chartsById": Object {},
  "nextId": 1,
}
`;
